### PR TITLE
Add audit annotations and metrics to track secret-based long-lived tokens

### PR DESCRIPTION
**Pull request by yuanchen8911**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
Track auto-generated and manually created secret-based long lived tokens.

1.   Add two audit annotations to track auto-generated and manually created long-lived tokens respectively.
  - `authentication.k8s.io/legacy-token-autogenerated-secret: <service account name> uses auto generated secret-based long lived token: <secret name>`
 - `authentication.k8s.io/legacy-token-manual-secret: <service account name> uses manually created secret-based long lived token: <secret name>`

2. Add two metrics  to count the total number of auto-generated and manually created tokens respectively.
- `auto_generated_tokens_total` 
-  `manually_created_tokens_total`

The enhancement will facilitate and encourage the migration from secret-based long-lived tokens to time-bound tokens  using TokenRequest API. 

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:
Manually test and validate the changes. 

#### Does this PR introduce a user-facing change?

```release-note
Use of secret-based service account tokens now adds an `authentication.k8s.io/legacy-token-autogenerated-secret` or `authentication.k8s.io/legacy-token-manual-secret` audit annotation containing the name of the secret used.
```
#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

Related work
[KEP]: [KEP-2799: Reduction of Secret-based Service Account Tokens ](https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/2799-reduction-of-secret-based-service-account-token)
[PR]: https://github.com/kubernetes/kubernetes/pull/108858

```docs

```

### CODE REVIEW

File: pkg/serviceaccount/legacy.go
Comment (liggitt) [inline pkg/serviceaccount/legacy.go:159]: I'm not sure I agree with warning here... this means someone is explicitly using a manually created token... we shouldn't add warnings unless we're 100% sure the caller can take action to resolve the warning, and they may not be able to update the caller to use refreshable/time-bound tokens

Comment (yuanchen8911) [reply pkg/serviceaccount/legacy.go:159]: We have discovered it is beneficial to categorize tokens into different types: automated generated, manually created, and obtained through TokenRequest APIs. Instead of adding a warning in the response and audit log, would it make sense to add extra logging for these different token types to the apiserver log? Thanks!

Comment (yuanchen8911) [reply pkg/serviceaccount/legacy.go:159]: We have discovered it is beneficial to categorize tokens into different types: automated generated, manually created, and obtained through TokenRequest APIs. Instead of adding a warning in the response and audit log, would it make sense to add extra logging for these different token types to the apiserver log? Thanks!

Comment (yuanchen8911) [reply pkg/serviceaccount/legacy.go:159]: In practice, we noticed quite a few (old) external systems, e.g., a CI/CD pipeline, store and use secret-based tokens to authenticate with Kubernetes. With such a warning, we hope the user could call `TokenRequest` API to fetch time bound tokens instead.

Comment (liggitt) [reply pkg/serviceaccount/legacy.go:159]: I'd be more open to decorating the audit log for use of a legacy token than returning a user-facing warning

Comment (yuanchen8911) [reply pkg/serviceaccount/legacy.go:159]: > I'd be more open to decorating the audit log for use of a legacy token than returning a user-facing warning

Thank, Jordan. Could you provide your recommendation on how to enhance the audit log for manually created tokens? I'd happy to update the PR.

Comment (liggitt) [reply pkg/serviceaccount/legacy.go:159]: you can follow the example of the information we add when using a stale time-bound token: https://github.com/kubernetes/kubernetes/blob/master/pkg/serviceaccount/claims.go#L208

Comment (yuanchen8911) [reply pkg/serviceaccount/legacy.go]: Hi Jordan, I reverted the changes to` legacy.go`. The updated PR tracks manually created tokens and adds extra info. to the audit log in `claim.go` instead as you suggested. Hopefully I did it right. Can you take a look? Thanks!

I'm not sure about how to test the change to the audit log. I haven't found a test for state tokens yet.

Comment (liggitt) [reply pkg/serviceaccount/legacy.go]: sorry, I wasn't suggesting to add it to claims.go, I was linking to that as an example of how to add audit annotations

Comment (yuanchen8911) [reply pkg/serviceaccount/legacy.go]: > sorry, I wasn't suggesting to add it to claims.go, I was linking to that as an example of how to add audit annotations

Sorry for misunderstanding it.  How about adding it to `legacy.go`  after the warning for auto-generated tokens? We can  add audit annotations for both auto-generated and manually-created tokens?  Thanks.

Comment (yuanchen8911) [reply pkg/serviceaccount/legacy.go]: Hi @liggitt, I've updated `legacy.go` to add audit annotations and metrics for auto-generated and manually created tokens. Can you please take a look when you get a chance? Thanks!

Comment (yuanchen8911) [reply pkg/serviceaccount/legacy.go]: @liggitt , I've made the suggested changes. Could you review it when you have a chance? Thanks!

Comment (yuanchen8911) [reply pkg/serviceaccount/legacy.go]: Hi Jordan, any feedback would be greatly appreciated. Thanks!

**Comment by yuanchen8911** [comment]
> /triage accepted

Thanks, @aramase! The PR has been pending for some time. It would be great if we can move it forward. Thanks!

File: pkg/serviceaccount/legacy.go
Comment (liggitt) [inline pkg/serviceaccount/legacy.go:157]: we already have the service account info in the audit log (it's the authenticated user), just use the name of the secret as the audit data, no need for a long human-readable message

```suggestion
				audit.AddAuditAnnotation(ctx, "authentication.k8s.io/legacy-token-autogenerated-secret", secret.Name)
```

File: pkg/serviceaccount/legacy.go
Comment (liggitt) [inline pkg/serviceaccount/legacy.go:166]:
```suggestion
			audit.AddAuditAnnotation(ctx, "authentication.k8s.io/legacy-token-manual-secret", secret.Name)
```

File: pkg/serviceaccount/legacy.go
Comment (yuanchen8911) [reply pkg/serviceaccount/legacy.go]: Thanks, Jordan! I've committed the suggested changes. Can you take another look please?

**Comment by liggitt** [comment]
this looks reasonable to me, it would be good to exercise this functionality with a test

**Comment by yuanchen8911** [comment]
> this looks reasonable to me, it would be good to exercise this functionality with a test

I tested it in a KinD cluster locally with auto-generated secrets and manual secretes. Both annotated the audit log correctly.

1. `authentication.k8s.io/legacy-token-manual-secret`
```
./kubernetes/kube-apiserver-audit.log:2719:

{"kind":"Status","apiVersion":"v1","metadata":"requestReceivedTimestamp":"2023-07-26T05:48:47.025015Z","stageTimestamp":"2023-07-26T05:48:47.028607Z","annotations":{"authentication.k8s.io/legacy-token":"system:serviceaccount:default:test-sa","authentication.k8s.io/legacy-token-manual-secret":"test-secret","authorization.k8s.io/decision":"allow","authorization.k8s.io/reason":"RBAC: allowed by RoleBinding \"pod-viewer/default\" of Role \"pod-viewer\" to ServiceAccount \"test-sa/default\""}}
```

2. `authentication.k8s.io/legacy-token-autogenerated-secret`
```
./kubernetes/kube-apiserver-audit.log:44615

,"requestReceivedTimestamp":"2023-07-26T04:35:38.662373Z","stageTimestamp":"2023-07-26T04:35:38.676176Z","annotations":{"authentication.k8s.io/legacy-token":"system:serviceaccount:default:default","authentication.k8s.io/legacy-token-autogenerated-secret":"default-token-nwq6w","authorization.k8s.io/decision":"allow","authorization.k8s.io/reason":"RBAC: allowed by RoleBinding \"pod-viewer/default\" of Role \"pod-viewer\" to ServiceAccount \"default/default\""}}
```

**Comment by dashpole** [comment]
/assign @dgrisonnet
to review the new metric

**Review by logicalhan — APPROVED** [review]
/lgtm
/approve

**Comment by yuanchen8911** [comment]
> /lgtm
> /approve

Hi @liggitt, sig-instrumentation has approved the PR. Does it look good to you too? Thanks!

**Comment by liggitt** [comment]
/lgtm
/approve

This will merge once master reopens for 1.29. Once it merges, please queue up a PR to the dev-1.29 branch of the kubernetes/website repo documenting the audit annotations

**Comment by yuanchen8911** [comment]
> /lgtm /approve
> 
> This will merge once master reopens for 1.29. Once it merges, please queue up a PR to the dev-1.29 branch of the kubernetes/website repo documenting the audit annotations

Thanks, will do it.

**Pull request merged**
